### PR TITLE
feat: Add support for gitops-bridge

### LINKED ISSUE #114: Improved integration between Terraform and GitOps providers
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

Better integration between Terraform and GitOps providers. There are infrastructure resources that are dynamically created by Terraform, and I wish to keep creating those resources in Terraform, but I need to pass their generated values (IDs, ARNs, etc.) over to my GitOps provider for downstream addons/applications to consume those values

#### Describe the solution you would like

In a perfect world, Terraform does not need to authenticate to the EKS cluster nor access inside the cluster. Terraform creates the necessary AWS infrastructure components, and passes those dynamically generated values over to my GitOps provider.

In addition, I would also like the ability to have the ability to choose between ArgoCD, FluxCD, or any other GitOps providers that may come along. To scope the current problem space - we will focus only on ArgoCD and FluxCD support.

#### Describe alternatives you have considered

None - this is an improvement over the current setup

#### Additional context

We often hear feedback from users that they wish to pass more values from Terraform to the GitOps provider via the associated `*_helm_config` variable. This request is tangentially related in that any dynamically generated values, as well as user provided values, can be passed over to the GitOps provider

- Reference #26
- Reference #40
- Reference #58
- Reference #59
- Reference #60
- Reference #104
- Reference  #119
- Reference #127 
- Reference https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1583
- Reference https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1579
- Reference https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1587
- Reference https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1609

### PR DESCRIPTION
**Pull request by csantanapr**
### What does this PR do?

Adds the ability to skip the use of terraform helm and kubernetes providers via new variable `create_kubernetes_resources`

Adds a new output `gitops_metadata` this output will have the aws metadata from the aws resources created in terraform, this metadata is to be use with a gitops engine such as ArgoCD or FluxCD to pass this metadata over a gitops-bridge.

To see example how this output is used checkout the gitops-bridge [terraform/examples/eks/](https://github.com/gitops-bridge-dev/gitops-bridge/tree/main/argocd/iac/terraform/examples/eks)

### Motivation

Resolves #114

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Yes, I ran `pre-commit run -a` with this PR

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -565,3 +565,20 @@ variable "aws_gateway_api_controller" {
   type        = any
   default     = {}
 }
+
+################################################################################
+# GitOps Bridge
+################################################################################
+
+# Only required when enable_aws_gateway_api_controller is true
+variable "vpc_id" {
```
Comment (csantanapr) [inline variables.tf:574]: @bryantbiggs what do you think about this new variable?
It makes a cleaner interface for the gitops bridge metadata.

It allows something like this
```hcl
module "gitops_bridge_metadata" {
  source = "../../../modules/gitops-bridge-metadata"

  cluster_name = module.eks.cluster_name
  metadata = module.eks_blueprints_addons.gitops_metadata
  environment = local.environment
  addons = local.addons
}
```
If `metadata_aws_vpc_id` doesn't come from `module.eks_blueprints_addons.gitops_metadata`
then it looks like this
```hcl
module "gitops_bridge_metadata" {
  source = "../../../modules/gitops-bridge-metadata"

  cluster_name = module.eks.cluster_name
  metadata = merge(module.eks_blueprints_addons.gitops_metadata, {
     metadata_aws_vpc_id = module.vpc.vpc_id
  })
  environment = local.environment
  addons = local.addons
}
```
I'm ok either way, we only need to doc if doing api gateway controller to pass the extra vpcid, if we leave it we could pass the vpc id into the helm_release when installing helm chart using terraform provider

Comment (bryantbiggs) [reply variables.tf:574]: apologies, this one slipped off my radar. In short - we should avoid passing through values and only export values created by this module. I'll take a deeper look at this PR tomorrow (8/9)

File: outputs.tf [hunk 2]
```diff
@@ -161,3 +161,102 @@ output "helm_releases" {
   description = "Map of attributes of the Helm release created"
   value       = helm_release.this
 }
+
+################################################################################
+# GitOps Bridge
+################################################################################
+
+output "gitops_metadata" {
```
Comment (bryantbiggs) [inline outputs.tf:169]: Two parts on this:
1. Lets not pass through - only export what is created here
2. Don't make users check for values later; if the resource isn't created, then don't export anything about it. Take a look at below:

```hcl
output "gitops_metadata" {
  description = "GitOps Bridge metadata"
  value = merge(
    { for k, v in {
      iamRoleArn     = module.aws_cloudwatch_metrics.iam_role_arn
      namespace      = local.aws_cloudwatch_metrics_namespace
      serviceAccount = local.aws_cloudwatch_metrics_service_account
      } : "awsCloudwatchMetrics_${k}" => v if var.enable_aws_cloudwatch_metrics
    },
    { for k, v in {
      iamRoleArn               = module.aws_efs_csi_driver.iam_role_arn
      namespace                = local.aws_efs_csi_driver_namespace
      controllerServiceAccount = local.aws_efs_csi_driver_controller_service_account
      nodeServiceAccount       = local.aws_efs_csi_driver_node_service_account
      } : "awsEfsCsiDriver_${k}" => v if var.enable_aws_efs_csi_driver
    },
    { for k, v in {
      iamRoleArn     = module.aws_for_fluentbit.iam_role_arn
      namespace      = local.aws_for_fluentbit_namespace
      serviceAccount = local.aws_for_fluentbit_service_account
      logGroupName   = try(aws_cloudwatch_log_group.aws_for_fluentbit[0].name, "")
      } : "awsForFluentbit_${k}" => v if var.enable_aws_for_fluentbit
    },
    { for k, v in {
      iamRoleArn               = module.aws_fsx_csi_driver.iam_role_arn
      namespace                = local.aws_fsx_csi_driver_namespace
      controllerServiceAccount = local.aws_fsx_csi_driver_controller_service_account
      nodeServiceAccount       = local.aws_fsx_csi_driver_node_service_account
      } : "awsFsxCsiDriver_${k}" => v if var.enable_aws_fsx_csi_driver
    },
    { for k, v in {
      iamRoleArn     = module.aws_gateway_api_controller.iam_role_arn
      namespace      = local.aws_gateway_api_controller_namespace
      serviceAccount = local.aws_gateway_api_controller_service_account
      } : "awsGatewayApiController_${k}" => v if var.enable_aws_gateway_api_controller
    },
    { for k, v in {
      iamRoleArn     = module.aws_load_balancer_controller.iam_role_arn
      namespace      = local.aws_load_balancer_controller_namespace
      serviceAccount = local.aws_load_balancer_controller_service_account
      } : "awsLoadBalancerController_${k}" => v if var.enable_aws_load_balancer_controller
    },
    { for k, v in {
      iamRoleArn     = module.aws_node_termination_handler.iam_role_arn
      namespace      = local.aws_node_termination_handler_namespace
      serviceAccount = local.aws_node_termination_handler_service_account
      sqsQueueUrl    = module.aws_node_termination_handler_sqs.queue_url
      } : "awsNodeTerminationHandler_${k}" => v if var.enable_aws_node_termination_handler
    },
    { for k, v in {
      iamRoleArn     = module.aws_privateca_issuer.iam_role_arn
      namespace      = local.aws_privateca_issuer_namespace
      serviceAccount = local.aws_privateca_issuer_service_account
      } : "awsPrivateCaIssuer_${k}" => v if var.enable_aws_privateca_issuer
    },
    { for k, v in {
      iamRoleArn     = module.cert_manager.iam_role_arn
      namespace      = local.cert_manager_namespace
      serviceAccount = local.cert_manager_service_account
      } : "certManager_${k}" => v if var.enable_cert_manager
    },
    { for k, v in {
      iamRoleArn     = module.cluster_autoscaler.iam_role_arn
      namespace      = local.cluster_autoscaler_namespace
      serviceAccount = local.cluster_autoscaler_service_account
      } : "clusterAutoscaler_${k}" => v if var.enable_cluster_autoscaler
    },
    { for k, v in {
      iamRoleArn     = module.external_dns.iam_role_arn
      namespace      = local.external_dns_namespace
      serviceAccount = local.external_dns_service_account
      } : "externalDns_${k}" => v if var.enable_external_dns
    },
    { for k, v in {
      iamRoleArn     = module.external_secrets.iam_role_arn
      namespace      = local.external_secrets_namespace
      serviceAccount = local.external_secrets_service_account
      } : "externalSecrets_${k}" => v if var.enable_external_secrets
    },
    { for k, v in {
      iamRoleArn              = module.karpenter.iam_role_arn
      namespace               = local.karpenter_namespace
      serviceAccount          = local.karpenter_service_account_name
      sqsQueueName            = module.karpenter_sqs.queue_name
      nodeInstanceProfileName = local.karpenter_node_instance_profile_name
      } : "karpenter_${k}" => v if var.enable_karpenter
    },
    { for k, v in {
      iam_role_arn    = module.velero.iam_role_arn
      namespace       = local.velero_namespace
      service_account = local.velero_service_account
      } : "velero_${k}" => v if var.enable_velero
    }
  )
}
```

File: outputs.tf [hunk 3]
```diff
@@ -161,3 +161,102 @@ output "helm_releases" {
   description = "Map of attributes of the Helm release created"
   value       = helm_release.this
 }
+
+################################################################################
+# GitOps Bridge
```
Comment (bryantbiggs) [inline outputs.tf:166]: lets add some more comments on this to make it clear why this is here, what its used for, our intentions, and our gurantees

File: outputs.tf (hunk 2 above)
Comment (bryantbiggs) [reply outputs.tf:169]: and I think below is the form that should work for https://github.com/gitops-bridge-dev/gitops-bridge/blob/main/argocd/iac/terraform/modules/gitops-bridge-metadata/main.tf (I had a local copy when I was playing around, but didn't plug back into the actual thing to test)

```hcl
locals {
  labels = {
    cluster_name  = var.cluster_name,
    environment   = var.environment
    enable_argocd = true
  }

  annotations = merge(
    local.labels,
    var.metadata,
    var.addons
  )

  argocd = {
    apiVersion = "v1"
    kind       = "Secret"
    metadata = {
      name        = try(var.argocd.secret_name, var.cluster_name)
      namespace   = try(var.argocd.secret_namespace, "argocd")
      annotations = { for key, value in local.annotations : key => tostring(value) if value != null }
      labels = merge(
        local.labels,
        { "argocd.argoproj.io/secret-type" = "cluster" }
      )

    }
    stringData = {
      name   = var.cluster_name
      server = try(var.argocd.server, "https://kubernetes.default.svc")
      config = try(var.argocd.argocd_server_config, {
        tlsClientConfig = {
          insecure = false
        }
      })
    }
  }

  fluxcd = {
    apiVersion = "v1"
    kind       = "ConfigMap"
    metadata = {
      name      = try(var.fluxcd.configmap_name, var.cluster_name)
      namespace = try(var.fluxcd.configmap_namespace, "flux-system")
    }
    data = { for key, value in local.cluster_config : key => tostring(value) if value != null }
  }
}
```

Comment (csantanapr) [reply outputs.tf:169]: Thanks for the review I will look into it and refactor, I would keep the snake case for the output keys

File: outputs.tf (hunk 3 above)
Comment (csantanapr) [reply outputs.tf:166]: I added comment here

File: variables.tf (hunk 1 above)
Comment (csantanapr) [reply variables.tf]: I remove the variable vpc_id and updated the examples to pass it to the metadata module

File: outputs.tf (hunk 2 above)
Comment (csantanapr) [reply outputs.tf:174]: @bryantbiggs I refactored the output, the metadata module, and all the examples, please take a look when you have a chance

**Comment by csantanapr** [comment]
@bryantbiggs I updated the PR to address the feedback, please take another look when you have chance

File: outputs.tf [hunk 4]
```diff
@@ -161,3 +161,122 @@ output "helm_releases" {
...
+      service_account = local.aws_node_termination_handler_service_account
+      sqs_queue_url   = module.aws_node_termination_handler_sqs.queue_url
+      } : "aws_node_termination_handler_${k}" => v if var.enable_aws_node_termination_handler
+    },
+    { for k, v in {
+      iam_role_arn               = module.karpenter.iam_role_arn
+      namespace                  = local.karpenter_namespace
+      service_account            = local.karpenter_service_account_name
+      sqs_queue_name             = module.karpenter_sqs.queue_name
+      cluster_endpoint           = local.cluster_endpoint
```
Comment (bryantbiggs) [inline outputs.tf:258]: remove this one too
```suggestion
```

File: outputs.tf [hunk 5]
```diff
@@ -161,3 +161,122 @@ output "helm_releases" {
...
+      node_instance_profile_name = local.karpenter_node_instance_profile_name
+      } : "karpenter_${k}" => v if var.enable_karpenter
+    },
+    { for k, v in {
+      iam_role_arn            = module.velero.iam_role_arn
+      namespace               = local.velero_namespace
+      service_account         = local.velero_service_account
+      backup_s3_bucket_prefix = local.velero_backup_s3_bucket_prefix
```
Comment (bryantbiggs) [inline outputs.tf:266]: remove both of these bucket attributes

File: outputs.tf [hunk 6]
```diff
@@ -161,3 +161,122 @@ output "helm_releases" {
...
+    },
+    { for k, v in {
+      iam_role_arn    = module.aws_gateway_api_controller.iam_role_arn
+      namespace       = local.aws_gateway_api_controller_namespace
+      service_account = local.aws_gateway_api_controller_service_account
+      } : "aws_gateway_api_controller_${k}" => v if var.enable_aws_gateway_api_controller
+    },
+    { for k, v in {
+      group_name    = try(var.fargate_fluentbit.cwlog_group, aws_cloudwatch_log_group.fargate_fluentbit[0].name, null)
+      stream_prefix = local.fargate_fluentbit_cwlog_stream_prefix
```
Comment (bryantbiggs) [inline outputs.tf:278]: remove this one too:
```suggestion
```

**Comment by csantanapr** [comment]
Hi @bryantbiggs I made the updates based on your feedback please another look when you have a chance

I refactored all the examples to meet the changes on the interface
https://github.com/gitops-bridge-dev/gitops-bridge/blob/main/argocd/iac/terraform/examples/eks/hello-world/main.tf#L85-L95
The mapping for cluster-autoscaler image tag I'm doing it in ArgoCD yaml https://github.com/gitops-bridge-dev/gitops-bridge-argocd-control-plane-template/blob/main/bootstrap/control-plane/addons/aws/addons-aws-oss-cluster-autoscaler-appset.yaml#L41-L94

**Review by bryantbiggs — APPROVED** [review]
Lets do it!

**Comment by csantanapr** [comment]
@bryantbiggs and  thank you for the review and merging 🍾

**Comment by bryantbiggs** [comment]
Thank you! Now let's get folks to quit sending Terraform into the cluster!!!

**Pull request merged**